File: lib/to_mdx/parsers/open_document/table/table_row.rb
# frozen_string_literal: true

require_relative "../base"
require_relative "table_cell"

module ToMdx
  module Parsers
    module Table
      class TableRow < Base
        def initialize(element = nil, context = {})
          super(element, context)
          @style_name = context[:styles].use(element&.attributes&.[]("table:style-name"))
        end

        def to_mdx
          content = children.map(&:to_mdx).join
          style_class = @style_name.empty? ? "" : " className=\"#{@style_name}\""
          "<tr#{style_class}>#{content}</tr>"
        end

        def to_s
          children.map(&:to_s).join(" | ")
        end

        register_for "table:table-row"
      end
    end
  end
end

File: src/parsers/open-document/table/table-row.ts
import dedent from "dedent-js";
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

declare module "../../base_element.ts" {
  interface ParseContext {
    defaultCellStyleName?: string | null;
  }
}

export class TableRow extends BaseElement {
  private styleName: Style;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.styleName = context.styles.use(this.attr("table:style-name"));
  }

  toMdx(): string {
    const content = this.children.map((child) => child.toMdx()).join("");
    const styleClass = this.styleName.empty() ? "" : ` className="${this.styleName}"`;
    return dedent`
      <tr${styleClass}>
        ${content}
      </tr>
    `;
  }

  toString(): string {
    return this.children.map((child) => child.toString()).join(" | ");
  }

  withContext(context: ParseContext): ParseContext {
    return {
      ...context,
      defaultCellStyleName: this.attr("table:default-cell-style-name"),
    };
  }

  static {
    BaseElement.registerFor.call(this, "table:table-row");
  }
}
